### DISCUSSION TITLE
Should we make ZITADEL APIs more easy to use?!

### DISCUSSION THREAD

**Discussion by fforootd** [zitadel/zitadel#4987] [Ideas]
Lets discuss how we can make the ZITADEL APIs the best and easiest to use!

## History

Since we started ZITADEL we always were committed to gRPC as API and used it for close to everything we built. To ease the pain with Single Page Apps we relied on [gRPC-Web](https://github.com/grpc/grpc-web) and [gRPC-Gateway](https://github.com/grpc-ecosystem/grpc-gateway) a lot. But it grew on me that the learning curve and overhead to maintain that construct might become an issue at one point.

## Goals

- We make our APIs lean and easy to understand
- We make our APIs really easy to test and integrate
- We document our APIs in a consistent way
- We ask the community about their opinion on this subject 💬 (please feel free to comment your opinion on this subject)

## The Challenges

- gRPC Tooling Accessibility - Even with the rise of [buf](https://buf.build/) the gRPC tooling always feels like an impediment 
- gRPC API Accessibility - While we use grpc-gateway its only patch to provide our users with a simple rest/openapi that can provide simple curl example in our docs... which I personally find critical for a great learning curve 😏 
- Ease of use - Our API architecture could require a rework to make it more easy to understand 😁 especially the seperation we made from auth/managment/... creates many redundancies
- API Documentation Quality - We definitely need to improve our API documentation 😉 

## Observation of "great APIs"

What should a great API bring along

- Easy and consistent documentation style
- Easy to test and integrate (should be usable with curl)
- Some kind of contract needs to be the basis for it (e.g. OpenAPI)

I think [stripe's API](https://stripe.com/docs/api) is a good example of a well designed an consistent API.

## Solutions

1. Create a new OpenAPI - This would definitely solve the accessibility part <- my personal favourite (even though some things needed to be discussed e.g code generator for Go/Angular)
2. Create a new streamlined gRPC API - This however does not solve the challenges around the accessibility
3. Create a new GraphQL API - I think this would be fancy but would not allow users to get easy started with our API

## Influencing Issues / PR / Discussions

- https://github.com/zitadel/zitadel/discussions/4193
- https://github.com/zitadel/zitadel/pull/4755
- https://github.com/zitadel/zitadel/discussions/4915
- https://github.com/zitadel/zitadel/issues/4839

**Comment by fforootd** [comment]
Whats your take on this @zitadel/engineers @dakshitha ?

**Comment by adlerhurst** [comment]
My opinion is to switch to OpenAPI (solution 1)

Considerations:

- Community adoption is more important than low latency
- The API is used to manage objects, not streaming or high traffic load
- Apps should use the oauth/saml API's for authN/authZ
- The API does not repleace SDK's

Pros:

- Many devs have heard or used some kind of REST based APIs
- Most programming languages support HTTP calls
- Native browser support (in comparisson grpc-web)
- HTTP/1.1 support
- Usage of HTTP status codes
- Devs can simply start without figuring out how to generate the client stub
  - No code generation required
- IMO: OpenAPI spec feels better organized
- JSON/YAML is more common to read than proto
- UI support thanks to Swagger UI

Cons:

- Slower than GRPC
- Lot of work to do for the change
- We have to change our mind for writing good API definition

**Comment by fforootd** [comment]
Some additional thoughts to this topic:

While talking to our customers there was actually an opinion coming up.

Why do we not provide SDKs for major languages/framework that are generated out of the current API and include the most common operations like authn, authz, User, Org, Management.

I think many devs would love to just import an SDK and get done with it.
Parallel to this we could better document the current API and reform it over time to become more easy to use. 
IMO we should merge all our APIs into one besides OIDC/SAML (which cannot be integrated)

**Reply by hifabienne** [reply]
I think the biggest problem in understanding our APIs at the moment is that we have splited them in multiple APIs. 
Not sure how easy it is to merge them together, instead of just creating a new one.
When we have decided on that, the next question is what we want to implement, openapi rest or grpc. 
The SDKs would anyway be great for the customers. But thats just on top of the APIs.

**Reply by eliobischof** [reply]
> I think the biggest problem in understanding our APIs at the moment is that we have splited them in multiple APIs. 
> Not sure how easy it is to merge them together, instead of just creating a new one.
> When we have decided on that, the next question is what we want to implement, openapi rest or grpc. 
> The SDKs would anyway be great for the customers. But thats just on top of the APIs.

Do you have feedback that says the division of the APIs is confusing? Because the APIs are now divided by caller groups like org admins, instance admins, system admins and authenticated users, right? My feeling says the categorization could help understanding their intent. 
Also, it's easy now to add a layer of security by locking down specific API groups by Firewalls or the like.

**Reply by fforootd** [reply]
Problem being that our internal knowledge is not applicable to external parties. 

At least I had feedback in that way. Because some functions felt redundant to users.

**Comment by peintnermax** [comment]
IMO we should offer PUT methods (REST or GRPC), which allow to modify the objects directly, without having to know the current state. 
Take our org policies as an example. They are designed to extend our instance policies, thus a GET should always return the values which are set on the instance if no explicit value is set on the org. And an update (PUT) method to the org should update the org state without having to know if there is a custom policy set or not.
REST APIs are IMO better suitable for managing objects and are easier to understand and to get started. REST could also mitigate the acute need for SDKs and, being more common, motivate more contributions.

**Comment by fforootd** [comment]
Ok so I had some time to think about this topic and my current view is as follows:

- Lets keep gRPC as API
  - In a first phase lets rework some calls (looking at a more easy way to create users, i.e having a simple call to update a user, this might need fieldmask) that do not work well and also rework some semantic errors (looking at the /users/{user_id} call)
  - In a second phase we should start separating resources in dedicated gRPC service (i.e user, org, instance, service and so on)
  - In a third phase we should leverage some of the streaming features of gRPC for fancy features 😁 
- Lets generate clients for all major frameworks / platforms
- Lets create a generated doc for proto with https://protobuffet.com/protodocs/protobuffet/example/ad/v1/ads.proto or something similar
- Lets also document the rest api in parallel (although i am not sure about the generation or redactional decision)

**Comment by hifabienne** [comment]
I just had some thoughts on this topic and I am not sure if we should work on new APIs for the moment. In general we did have reasons why we did structure the APIs like this.  (It would probably be easier to understand if we would only have one API, but I don't think this is the constraint right now)

I think were we definitely lack is the documentation and that should be improved, if we know change the api we will probably need way more time and the documentation will not be directly better, as I think it will be some manual work as it is not easy achievable with generated API docs.

So from a customer/developer perspective what I want is: Find out how to manage my user, how to manage some resources etc.
Proposal restructure API Docuementation:

Main Menu: API Reference
Sub Menu:
- Introduction (basically what we have in the overview)
- Manage my user (authentication API)
- Manage organization
   - Resource based structure
   - Settings
   - Users
   - Projects
   - etc.
- Manage ZITADEL
   - Resource based structure
   - Settings
   - Organisations
   - etc.
- Manage Virtual instance (note this is only for self hosting)

**Comment by cvvs** [comment]
It may be worth it to maintain both gRPC and ReST endpoints. There are tools which allow teams to generate OpenAPI v3 specs from protobufs, and at ${CURRENT_EMPLOYER}, we leverage this and Envoy proxy's ability to translate ReST calls to gRPC for our customers who don't use gRPC. Having used these in anger, they are not without their warts, but it's worth considering.

**Reply by fforootd** [reply]
Totally agree with your observation here.

At the moment we will keep the gRPC and Rest API but we might restructure it to better algin with our evolution so far.
I think tooling wise we need to make the life of people easier by publishing ready made clients and/or a buf registry 😁

**Comment by ghstahl** [comment]
Opinion: This is not an error
If you went to the trouble of telling me nothing changed, then I am ok with you saying it succeeded.  
It also makes infrastructure as code easier to write if adopting the pattern that the outcome is more important than how we got there. 

I would like to see this API design in zitadel.

Example of and error that probably shouldn't be.
```json
{
    "code": 9,
    "message": "Errors.Org.LoginPolicy.NotChanged (Org-5M9vdd)",
    "details": [
        {
            "@type": "type.googleapis.com/zitadel.v1.ErrorDetail",
            "id": "Org-5M9vdd",
            "message": "Errors.Org.LoginPolicy.NotChanged"
        }
    ]
}
```

**Reply by hifabienne** [reply]
Hei @ghstahl 
Thanks for your feedback.
You are absolutely right. We did implement all of the requests in this way, but we now changed that behaviour for new requests. 
E. g the new identity providers templates we have just built do not send this error anymore.
As it would be a breaking behaviour change, to just remove it from all requests we will not do that for the moment.

We are currently starting our new ZITADEL API, which will include the User Register and Login Endpoints in the first place. In the future the API should include all requests so we can get rid of the API separation regarding the context. 
And in the new API we will not have that error anymore.

**Reply by corepay** [reply]
Go look at scalar.com.  Sexy as hell API docs with a Notion-like editor.
